cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,922 @@
+    {
+      return error;
+    }
+
+    const int items_per_thread =
+      ::cuda::ceil_div(ActivePolicy::min_bif, max_occupancy * block_dim * loaded_bytes_per_iter);
+    const int items_per_thread_clamped =
+      ::cuda::std::clamp(items_per_thread, +policy_t::MIN_ITEMS_PER_THREAD, +policy_t::MAX_ITEMS_PER_THREAD);
+    const Offset tile_size = block_dim * items_per_thread_clamped;
+    const Offset grid_dim  = ::cuda::ceil_div(num_items, tile_size);
+
+    return THRUST_NS_QUALIFIER::cuda_cub::launcher::triple_chevron(grid_dim, block_dim, 0, stream)
+      .doit(kernel,
+            num_items,
+            items_per_thread_clamped,
+            op,
+            out,
+            THRUST_NS_QUALIFIER::unwrap_contiguous_iterator(::cuda::std::get<Is>(in))...);
+  }
+
+  template <std::size_t... Is>
+  struct non_contiguous_fallback_op_t
+  {
+    ::cuda::std::tuple<RandomAccessIteratorsIn...> in;
+    RandomAccessIteratorOut out;
+    mutable TransformOp op; // too many users forgot to mark there operator()'s const ...
+
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void operator()(Offset i) const
+    {
+      out[i] = op(::cuda::std::get<Is>(in)[i]...);
+    }
+  };
+
+  template <typename, std::size_t... Is>
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t
+  invoke_algorithm(cuda::std::index_sequence<Is...>, ::cuda::std::integral_constant<Algorithm, Algorithm::fallback_for>)
+  {
+    using op_t = non_contiguous_fallback_op_t<Is...>;
+    return for_each::dispatch_t<Offset, op_t>::dispatch(num_items, op_t{in, out, op}, stream);
+  }
+
+  // TODO(bgruber): check if there is any benefit in having a dedicated path for no input streams
+  // struct fill_opt_t
+  // {
+  //   RandomAccessIteratorOut out;
+  //   mutable TransformOp op; // too many users forgot to mark their operator()'s const ...
+  //
+  //   _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void operator()(Offset i) const
+  //   {
+  //     out[i] = op();
+  //   }
+  // };
+  //
+  // template <typename>
+  // CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t
+  // invoke_algorithm(cuda::std::index_sequence<>, ::cuda::std::integral_constant<Algorithm, Algorithm::fallback_for>)
+  // {
+  //   return for_each::dispatch_t<Offset, fill_opt_t>::dispatch(num_items, fill_opt_t{out, op}, stream);
+  // }
+
+  template <typename ActivePolicy>
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t Invoke()
+  {
+    constexpr bool all_contiguous =
+      ::cuda::std::conjunction<THRUST_NS_QUALIFIER::is_contiguous_iterator<RandomAccessIteratorsIn>...>::value;
+    constexpr bool no_input_streams = sizeof...(RandomAccessIteratorsIn) == 0;
+
+    // TODO(bgruber): transform_unrolled_staged_kernel would work for non-contiguous iterators. Try if it's faster.
+    // TODO(bgruber): use an if constexpr here in C++17
+    constexpr Algorithm algo =
+      no_input_streams || !all_contiguous ? Algorithm::fallback_for
+      : CanCopy                           ? ActivePolicy::alg_addr_unstable
```

Comment:
question: today we compile one kernel per arch. After this change, we'll compile three. This affects both compilation time and binary sizes. Can we do anything to get back to one kernel per arch?

Reply:
I don't completely follow. We always compile 1 kernel per instantiation of `thrust::transform`. Which one depends on whether the iterators are contiguous and whether I can copy the data before passing it to the lambda.

Reply:
This is a quite insidious part of the CUB dispatch mechanism. I will try to explain why we would instantiate three different kernels here, even if we would only need to compile one.

So, during compilation, CUB's dispatch mechanism (more specifically, `cub::ChainedPolicy`) instantiates the `Invoke` function template for all tunings policies that are defined in the algorithm's chain of tuning policies. 

At runtime, when an algorithm is invoked, of all the instantiated `Invoke`-function templates, `cub::ChainedPolicy` makes sure to call only the first tuning policy whose PTX version is less-than-or-equal to the GPU's SM version to which the algorithm is dispatched.

### Example:

Let's assume `(no_input_streams || !all_contiguous)` is `false` and `CanCopy` is `true`. Or, in short: We will always choose the `ActivePolicy::alg_addr_unstable` code path. Now, what is `ActivePolicy::alg_addr_unstable`? That depends on `ActivePolicy`:
- `policy300`: `Algorithm::unrolled_staged` ▶️ `transform_unrolled_staged_kernel`
- `policy800`: `Algorithm::memcpy_async;`▶️ `transform_memcpy_async_kernel`
- `policy900`: `Algorithm::ublkcp;`▶️ `transform_ublkcp_kernel`

Since the `Invoke` function template gets instantiated for each of these tunings policies, we now have:
- `Invoke<policy300>`: `transform_unrolled_staged_kernel`
- `Invoke<policy800>`: `transform_memcpy_async_kernel`
- `Invoke<policy900>`: `transform_ublkcp_kernel`

And the logic in `cub::ChainedPolicy` will compile to what would resemble something like:

```
device_ptx_version=PtxVersion(); // query closest ptx version we had compiled for for the current device
if (device_ptx_version >= 900)
{
    dispatch_op.template Invoke<policy900>(); // calls transform_ublkcp_kernel<<<>>>();
}
if (device_ptx_version >= 800)
{
    dispatch_op.template Invoke<policy800>(); // calls transform_memcpy_async_kernel<<<>>>();
}
if (device_ptx_version >= 300)
{
    dispatch_op.template Invoke<policy800>(); // calls transform_unrolled_staged_kernel<<<>>>();
}
```

Since the `Invoke` function template is instantiated for all tunings policies in the chain, we want to avoid (1) making any of the algorithm's kernel template parameters depend on the `ActivePolicyT` template argument of the `Invoke` function template and (2) making the kernel we invoke depend on `ActivePolicyT` within the `Invoke` function template. (1) here is also the reason why we pass in the `MaxPolicy` to the kernel and not the `ActivePolicy` ([here's an instance of where this was problematic before](https://github.com/NVIDIA/cub/pull/312)). 

I yet need to look into the details of whether we can actually even resolve - or even improve - this here. It's quite challenging given we have quite different code paths. Generally speaking, as a resolution, we would want to have _one single_ kernel that we launch for any "algorithm" specialization, and its implementation will query `ActivePolicy::alg_addr_unstable` and, e.g., tag-dispatch to the according specialization.

Reply:
Aside: I tried to prune the number of policies we compile for on the host side when `__CUDA_ARCH_LIST__` is available: #2154

Reply:
As a fallback for `no_input_streams || !all_contiguous` I am using `Algorithm::prefetch` now, which I extended to support any iterators now (prefetching a non-contiguous iterator will just do nothing now).

Reply:
I rewrote the entire dispatching logic now. The policy hub decides which algorithm to use (stable vs. unstable address, fallback, etc.), making the policies themselves simpler. I now have a single kernel that is called from each implementation strategy with the same template arguments, but potentially different values. The cursed ingredient is a `union` to erase the type used by the respective implementation path on the device side.

I compiled with `-DCMAKE_CUDA_ARCHITECTURES=86;90` and then ran `cuobjdump -sass bin/cub.test.device_transform.lid_0 | grep plus_needs_stable_address` (see the unit test calling transform a functor called `plus_needs_stable_address`) and I got exactly two hits. One kernel per architecture we compile for.

I hope you are happy now, because I cried a whole lot when writing this code! ;)

Reply:
@gevtushenko and @elstehle please have another look whether the new implementation is satisfactory!

Reply:
Ping.

Reply:
> I hope you are happy now, because I cried a whole lot when writing this code! ;)

@bernhardmgruber you've come up with a clever solution! I think it works.

Reply:
Hand-made type erasured union through a byte array == clever solution. Alright. I think I understand the spirit here :D :D